runtime/runtime-core/common/test/aws/smithy/kotlin/runtime/content/ByteStreamFlowTest.kt

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.content
+
+import io.kotest.matchers.string.shouldContain
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.test.runTest
+import java.lang.RuntimeException
+import kotlin.test.*
+
+class ByteStreamBufferFlowTest : ByteStreamFlowTest(ByteStreamFactory.BYTE_ARRAY)
+class ByteStreamSourceStreamFlowTest : ByteStreamFlowTest(ByteStreamFactory.SDK_SOURCE)
+class ByteStreamChannelSourceFlowTest : ByteStreamFlowTest(ByteStreamFactory.SDK_CHANNEL)
+
+abstract class ByteStreamFlowTest(
+    private val factory: ByteStreamFactory,
+) {
+    @Test
+    fun testToFlowWithSizeHint() = runTest {
+        val data = "a korf is a tiger".repeat(1024).encodeToByteArray()
+        val bufferSize = 8182 * 2
+        val byteStream = factory.byteStream(data)
+        val flow = byteStream.toFlow(bufferSize.toLong())
+        val buffers = mutableListOf<ByteArray>()
+        flow.toList(buffers)
+
+        val totalCollected = buffers.fold(0) { acc, bytes -> acc + bytes.size }
+        assertEquals(data.size, totalCollected)
+
+        if (byteStream is ByteStream.Buffer) {
+            assertEquals(1, buffers.size)
+            assertContentEquals(data, buffers.first())
+        } else {
+            val expectedFullBuffers = data.size / bufferSize
+            for (i in 0 until expectedFullBuffers) {
+                val b = buffers[i]
+                val expected = data.sliceArray((i * bufferSize)until (i * bufferSize + bufferSize))
+                assertEquals(bufferSize, b.size)
+                assertContentEquals(expected, b)
+            }
+
+            val last = buffers.last()
+            val expected = data.sliceArray(((buffers.size - 1) * bufferSize) until data.size)
+            assertContentEquals(expected, last)
+        }
+    }
+
+    class FlowToByteStreamTest {
+        private fun randomByteArray(size: Int): ByteArray = ByteArray(size) { i -> i.toByte() }
```

Comment:
**Nit**: This isn't a "random" byte array. It's a "sequential" byte array or a "test" byte array.